[kotlin][rfc] J2K: Add pre- and post-processing extension points

ermattt opened the pull request:
Based on testing this extension point with two dummy preprocessors (see [draft PR](https://github.com/ermattt/intellij-community/pull/45/files)), this works as expected (i.e. it runs the preprocessors on a background thread, and successfully applies their changes before conversion) with _one major caveat_: changes due to conversion and changes due to preprocessing require two undo actions (i.e. Ctl + z).

I've tried a couple things to fix this problem, but so far none has worked:
1. Running the custom preprocessors under a transparent global command ([code](https://github.com/ermattt/intellij-community/pull/47/files)).  Maybe I was supposed to enable merging of global commands somehow?
2. Using the same command group id to run the custom preprocessors and the committing new kotlin files + external code processing command ([code](https://github.com/ermattt/intellij-community/pull/46/files))

You can see the example preprocessors in action (plus the suboptimal undo behavior) in the attached video zip.
[Screen Recording 2024-06-24 at 11.08.57 AM.zip](https://github.com/user-attachments/files/15959012/Screen.Recording.2024-06-24.at.11.08.57.AM.zip)



Once the issues with the preprocessing extension point are resolved, I'll add a postprocessing extension point by more or less copying this one.

[KTIJ-29835](https://youtrack.jetbrains.com/issue/KTIJ-29835/J2K-Add-extension-point-for-custom-conversions-processings)

@abelkov @darthorimar @jocelynluizzi13

darthorimar commented on plugins/kotlin/j2k/k1.new.post-processing/src/org/jetbrains/kotlin/idea/j2k/post/processing/utils.kt:
```diff
@@ -6,7 +6,7 @@ import com.intellij.openapi.application.ApplicationManager
 import com.intellij.openapi.command.CommandProcessor
 
-internal fun runUndoTransparentActionInEdt(inWriteAction: Boolean, action: () -> Unit) {
+fun runUndoTransparentActionInEdt(inWriteAction: Boolean, action: () -> Unit) {
```
This is an internal J2K function, let's not expose it :)

darthorimar commented on plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt:
```diff
@@ -0,0 +1,35 @@
+ */
+interface J2kPreprocessorExtension {
+
+    /**
+     * Override this method to analyze and edit Java files before conversion. This method is always executed on a background thread, so
+     * write actions must be wrapped in `runUndoTransparentActionInEdt(inWriteAction = true) { ... }`. As usual, read actions must be
+     * wrapped in `runReadAction { ... }`, and analysis must be done outside write actions.
+     */
```
Maybe in this case, it would be better to split `processFiles` into `processFiles` + `apply`? So the first one is executed from a BGT + read action while the second one is from the EDT + writeAction.

Another alternative would be to migrate this code to coroutines. There are `com.intellij.openapi.application.CoroutinesKt#writeAction` and `com.intellij.openapi.application.CoroutinesKt#readAction`. So the implementation of `processFiles` will look like

```
override suspend fun processFiles(project: Project, files: List<PsiJavaFile>) {`
  val result =   readAction { ... }
  writeAction { /* apply to the PSI tree here */ } // no explicit switching to EDT
}
```

ermattt replied:
Done!

darthorimar commented on plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/PreprocessorExtensionsRunner.kt:
```diff
@@ -0,0 +1,48 @@
+        if (preprocessorExtensions.isEmpty()) return
+
+        val preprocessorsCount = preprocessorExtensions.size
+        ProgressManager.progress(PHASE_NAME, "Found $preprocessorsCount preprocessors to run on Java files before conversion")
+
+        for ((i, preprocessor) in preprocessorExtensions.withIndex()) {
+            ProgressManager.checkCanceled()
+            ProgressManager.progress(PHASE_NAME, "Running preprocessor $i/$preprocessorsCount")
```
The indexation here will be from `0`

ermattt commented:
Tests added and coroutines funkiness resolved!
I probably need to clean up some k1/k2 test duplicates, but other than that I think this is ready for review again. If it looks good to you, @darthorimar, I'll expand upon these changes to handle postprocessing as well.

darthorimar commented:
@ermattt Awesome! Now tests behave more like production in `JavaToKotlinAction`.

The formatting changes in test data are definitely magic, but they seem to simulate the behavior in the real IntelliJ IDEA, so they are totally okay.

I'm ready to merge this PR as is, or do you want to have some modifications here?

>  If it looks good to you, @darthorimar, I'll expand upon these changes to handle postprocessing as well.
Definitely, thanks!

ermattt commented:
I'll add my postprocessing changes to this PR today, test 'em, and then re-request review from you 🥳

abelkov commented on plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:
```diff
@@ -138,26 +149,39 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
         val postProcessor = extension.createPostProcessor()
-        return converter.filesToKotlin(listOf(file), postProcessor).results.single()
+        var converterResult: FilesResult? = null
+        project.executeCommand("J2K") {
```
We need this machinery for coroutines to work, right?

ermattt replied:
Unfortunately yes

abelkov commented on plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:
```diff
@@ -138,26 +149,39 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
 
     private fun methodToKotlin(text: String, settings: ConverterSettings): String {
-        val result = fileToKotlin("final class C {$text}", settings)
+        val result = fileToKotlin("final class C {$text}", settings, preprocessorExtensions = emptyList())
```
Can be a default parameter value

abelkov commented on plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:
```diff
@@ -189,4 +213,21 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
     private fun createKotlinFile(text: String): KtFile =
         myFixture.configureByText("converterTestFile.kt", text) as KtFile
+
+    private val dummyPreprocessorEP = object : J2kPreprocessorExtension {
```
Probably worth to extract to a top-level J2KTestPreprocessorExtension

abelkov commented on plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt:
````diff
@@ -0,0 +1,39 @@
+ * </extensions>
+ * ```
+ *
+ * All preprocessors are run on a background thread, so write actions must be wrapped in
````
Probably worth to mention that they are run in coroutines and therefore must use coroutine-compatible API

abelkov commented on plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt:
````diff
@@ -0,0 +1,39 @@
+ * ```
+ *
+ * All preprocessors are run on a background thread, so write actions must be wrapped in
+ * `writeAction { ... }` so that they are executed on the EDT thread. As usual, read actions must
````
FQNs for writeAction/readAction here would be nice to avoid confusion

abelkov commented on plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:
```diff
@@ -189,4 +213,21 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
+        ) {
+            for (file in files) {
+                val method = readAction {
+                    file.classes.firstOrNull()?.findDescendantOfType<PsiMethod> { it.name != "main" && !it.isConstructor && !it.name.startsWith("get") && !it.name.startsWith("set") }
```
long line

abelkov commented on plugins/kotlin/j2k/shared/tests/testData/newJ2k/annotations/annotationUsages.k1.kt:
```diff
@@ -18,7 +18,8 @@ import javaApi.E
 @Anon6("x", "y")
 @Anon7(
-    String::class, StringBuilder::class
+    String::class,
```
Formatting changes are because we now run test J2K inside of a command? Does this indeed reflect production J2K results?

ermattt replied:
So it seems! I have no idea why, but it lines up with the results i get if I try converting a similar expression in a test instance of the IDE

ermattt commented on plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt:
```diff
@@ -0,0 +1,39 @@
+
+    /**
+     * Override this method to analyze and edit Java files before conversion. This method is always executed on a background thread, so
+     * write actions must be wrapped in `runUndoTransparentActionInEdt(inWriteAction = true) { ... }`. As usual, read actions must be
```
My bad, that was an outdated kdoc!

ermattt commented on plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/actions/JavaToKotlinAction.kt:
```diff
@@ -82,7 +79,12 @@ class JavaToKotlinAction : AnAction() {
+                        javaFiles,
+                        postProcessor,
+                        progressIndicator,
+                        preprocessorExtensions = J2kPreprocessorExtension.EP_NAME.extensionList
```
It's still too long for one line when I include postprocessors too :(

ermattt commented:
@abelkov & @darthorimar, addressed comments and tested the example pre- and post-processor extensions together (see screen recording)
[preAndPostTogether.mov.zip](https://github.com/user-attachments/files/16189086/preAndPostTogether.mov.zip)

abelkov commented on plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:
```diff
@@ -190,3 +222,45 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
+    }
+}
+
+private val dummyPostprocesorEP = object : J2kPostprocessorExtension {
```
Let's move them to testUtils.kt and make them internal named objects?

abelkov commented on plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/preAndPostprocessorExtensionsRunner.kt:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    private fun commitAllDocuments(javaFiles: List<T>) {
```
`files`

abelkov commented on plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:
```diff
@@ -138,22 +149,43 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
-        return converter.filesToKotlin(listOf(file), postProcessor).results.single()
+        var converterResult: FilesResult? = null
+        project.executeCommand("J2K") {
+            ProgressManager.getInstance()
```
A bit nicer formatting:

```
val process = {
    converterResult = converter.filesToKotlin(
        listOf(file),
        postProcessor,
        EmptyProgressIndicator(),
        preprocessorExtensions,
        postprocessorExtensions
    )
}
project.executeCommand("J2K") {
    ProgressManager.getInstance().runProcessWithProgressSynchronously(process, "Testing J2K", /* canBeCanceled = */ true, project)
}
```

ermattt commented:
@abelkov just addressed the last round of comments
